src/routers/alpha-router/alpha-router.ts

```diff
@@ -119,51 +119,85 @@ export class AlphaRouter implements IRouter<AlphaRouterConfig>, ISwapToRatio<Alp
-      const sqrtPriceX96Upper = TickMath.getSqrtRatioAtTick(position.tickUpper)
-
-      const token0Proportion = SqrtPriceMath.getAmount0Delta(
-        sqrtPriceX96,
-        sqrtPriceX96Upper,
-        precision,
-        true
-      )
-      const token1Proportion = SqrtPriceMath.getAmount1Delta(
-        sqrtPriceX96,
-        sqrtPriceX96Lower,
-        precision,
-        true
+      let optimalRatio = this.calculateOptimalRatio(
+        position,
+        position.pool.sqrtRatioX96
       )
 
-      const amountToSwap = calculateRatioAmountIn(
-        new Fraction(token0Proportion, token1Proportion),
-        position.pool.token0Price,
-        token0Balance,
-        token1Balance,
-      )
+      let zeroForOne: boolean
+      if (position.pool.tickCurrent < position.tickLower) {
+        zeroForOne = true
+      } else if (position.pool.tickCurrent > position.tickUpper) {
+        zeroForOne = false
+      } else {
+        zeroForOne = new Fraction(token0Balance.quotient, token1Balance.quotient).greaterThan(optimalRatio)
+        if (!zeroForOne) optimalRatio = optimalRatio.invert()
+      }
 
-      return this.routeExactIn(
-        amountToSwap.currency,
-        amountToSwap.currency == token0Balance.currency
-          ? token1Balance.currency
-          : token0Balance.currency,
-        amountToSwap,
-        swapConfig,
-        routingConfig
-      )
+      const [inputBalance, outputBalance] = zeroForOne
+        ? [token0Balance, token1Balance]
+        : [token1Balance, token0Balance]
+
+      let exchangeRate: Fraction = zeroForOne
+        ? position.pool.token0Price
+        : position.pool.token1Price
+      let swap: SwapRoute<TradeType.EXACT_INPUT> | null = null
+      let ratioAchieved = false
+
+      while (!ratioAchieved) {
+        let amountToSwap = calculateRatioAmountIn(
+          optimalRatio,
+          exchangeRate,
+          inputBalance,
+          outputBalance,
+        )
+
+        swap = await this.routeExactIn(
+          inputBalance.currency,
+          outputBalance.currency,
+          amountToSwap,
+          swapConfig,
+          routingConfig
+        )
+        if (!swap) {
+          throw new Error('did not fetch swap')
+        }
+
+        let inputBalanceUpdated = inputBalance.subtract(swap.trade.inputAmount)
+        let outputBalanceUpdated = outputBalance.add(swap.trade.outputAmount)
+        let newRatio = inputBalanceUpdated.divide(outputBalanceUpdated)
+
+        swap.route.forEach(route => {
+          route.route.pools.forEach((pool, i) => {
+            if(
+              pool.token0 == position.pool.token0 &&
+              pool.token1 == position.pool.token1 &&
+              pool.fee == position.pool.fee
+            ) {
+              optimalRatio = this.calculateOptimalRatio(
+                position,
+                JSBI.BigInt(route.sqrtPriceX96AfterList[i]!.toString())
+              )
```

Comment:
Hmm won't this cause issues for our next iteration? 

Lets say on our first iteration our route uses the pool we are adding to and we update the optimal ratio. Then on our second iteration we'll have a new target optimal ratio, but we aren't guaranteed that the next `routeExactIn` will use the pool we are adding too (or even send the same % amount of the input to that pool) so our target ratio would be off

Reply:
My thinking was that each iteration we essentially want to throw away the swap we just tried, reset to the state at the beginning of the block, and try a completely new swap.

If we use an updated sqrtPrice on the next iteration, we would be trying to find the ratio assuming we executed the swap from the previous iteration (which we wouldn't actually do).

Maybe I'm completely misunderstanding though, if so maybe we can hop on a call so you can set me right

Reply:
It's currently going off the assumption that the next swap will swap a similar amount therefore having a similar effect. NOT that the last swap already happened, only that our next try will *also* move the price in the same way.

If we get to a point where our output quote is really lining up with what we expected it's only that we budged the sqrtPrice hence shifting optimalRatio, we need to try again swapping just a little bit more to reach that adjusted optimalRatio, but we need to know what that new optimalRatio is to do that, hence account for the new price. If the price is no longer pushed when trying to hit the new optimalRatio, then that's potentially infinite loop territory where we go back and forth.

Reply:
I don't think we can assume that the next swap will go through the same pools though, so the sqrtPrice may not have moved at all. Changing the amount in a swap can result in very different route recommendations. This is the sort of example I'm worried about:

I have USDC. I want to add to USDC/USDT/5% pool

Iteration 1:

- `route.routeExactIn` recommends swap through the pool we are adding to (USDC —5%—>USDT )
- We use the new sqrtPrice of USDC/USDT/5% pool to compute optimalRatio. We use that to compare our ratio. We find it isn't close enough, so we proceed to next iteration.

Iteration 2:

- `route.routeExactIn` recommends a multi-hop through different pools (USDC —1%—>ETH—1%—>USDT )
- We use the optimalRatio from iteration 1 (which assumes we swapped on the USDC/USDT/5% pool, but we didn't) as our target ratio.
    - Therefore our target ratio is off?

Reply:
I guess I'm wondering if it should be like:
```
let newOptimalRatio = optimalRatio
swap.route.forEach(route => {
          route.route.pools.forEach((pool, i) => {
            if(
              pool.token0.equals(position.pool.token0) &&
              pool.token1.equals(position.pool.token1) &&
              pool.fee == position.pool.fee
            ) {
              newOptimalRatio = this.calculateOptimalRatio(
                position,
                JSBI.BigInt(route.sqrtPriceX96AfterList[i]!.toString())
              )
            }
          })
        })

        ratioAchieved = (
          newRatio.equalTo(newOptimalRatio) ||
          this.absoluteValue(newRatio.asFraction.divide(newOptimalRatio).subtract(1)).lessThan(errorTolerance)
```

Reply:
haha I'm not sure whether we are talking past each other or what! If we don't run the next iteration with newOptimalRatio, we'll never hit it, right? Because we keep aiming for optimalRatio all else being constant.

In your example above, I'd imagine there'd be an iteration3 that switches back to the previous `optimalRatio` and runs it again. The bad scenario here, is that the difference between optimalRatio and newOptimalRatio is what triggers the different trade path. My naive assumption is that getting the wrong exchange rate will mostly be determining that, esp in very liquid pools.

Reply:
I accidently deleted my comment that had the code so just reposting it:

```
let newOptimalRatio = optimalRatio
swap.route.forEach(route => {
          route.route.pools.forEach((pool, i) => {
            if(
              pool.token0.equals(position.pool.token0) &&
              pool.token1.equals(position.pool.token1) &&
              pool.fee == position.pool.fee
            ) {
              newOptimalRatio = this.calculateOptimalRatio(
                position,
                JSBI.BigInt(route.sqrtPriceX96AfterList[i]!.toString())
              )
            }
          })
        })

        ratioAchieved = this.absoluteValue(
          newRatio.asFraction.divide(newOptimalRatio).subtract(1)
        ).lessThan(swapAndAddConfig.errorTolerance)
        exchangeRate = swap.trade.outputAmount.divide(swap.trade.inputAmount)
```

Well if we don't run the next iteration with `newOptimalRatio`, our exchange rate would still have changed, so I'd expect the next iteration to still get us closer to the optimal ratio. 

It would be the same situation as if our first iteration didn't use the pool we are adding to, so optimalRatio ends up being the same for our second iteration.

But yeah maybe we are talking past each other! Might be good to get @NoahZinsmeister thoughts on it too

Reply:
so from my perspective, i think it's ok to update the `optimalRatio` according to the new market price of the target pool after the swap we've just tried. i think the obvious case where this makes sense is if increasing the amount swapped always routed the same or more volume through the target pool. if so, we know the `optimalRatio` is going to be at least what the post-swap price is (we all should be able to agree on this case).

however, even when that's not the case, i still think updating is correct. say we do a single round, the swap routes through our target pool, we update the `optimalRatio`, then do the next round (with a higher swap amount). if this round doesn't route through the same pool, we'll have overshot, but because we still update `exchangeRate`, we still have an accurate picture w/r/t the price impact of that trade size, so our next round (targeting the original `optimalRatio` with the new `exchangeRate`, should compensate for this overshoot, and we should be able to iterate toward the optimal solution.

the one important caveat here is that if a swap _doesn't_ use the target pool, we have to be sure to update the optimalRatio back to what it was before the loop, which afaict the current code doesn't do, it only updates it in the case when the pool is used, but then it's never reset. does that make sense?

Reply:
thanks, what you mention in your last paragraph is definitely a bug, I will fix that.

Reply:
Makes sense to me, though yeah currently the code doesn't reset the optimalRatio

Imo it would be simpler to just reset it on every iteration, but I suppose having the code where we don't reset initially could make it converge quicker in the case where the proposed routes always use the target pool, so I'm ok with it